### ISSUE TITLE
EhCache plugin is unable to discover new caches added at runtime through JCache API

### ISSUE THREAD

**Issue by yacota** [stagemonitor/stagemonitor#127]
It would be useful that EhCache plugin could discover new caches added at runtime.

By now EhCache plugin discovers a single CacheManager based on "stagemonitor.properties#stagemonitor.ehcache.name ", and through it discovers all the configured Cache instances.

Two chances here as far as I can see:
- let EhCachePlugin reload its configuration in order to be able to detect these new Cache instances.
- launch StageMonitor in lazy mode, avoiding the automatic initialization that WebPlugin triggers, which happens because it implements ServletContainerInitializer. So launching StageMonitor in lazy mode would imply that the  application itself should invoke StageMonitor.init() in order to initialize it.

To sum up, I am unable to configure a EhCacheCacheManager before StageMonitor is initialized, and only caches in ehcache.xml are detected.

Any thoughts?

**Comment by felixbarny** [comment]
One idea would be to offer a method that takes an `EhCacheCacheManager` and discovers the cache names from it. Could you describe how you add the caches at runtime? Why aren't they in ehcache.xml?

**Comment by yacota** [comment]
Why are not they listed in ehcache.xml ? The idea is to have a single common base configuration in ehcache.xml, which will be applied by default.  
At runtime/startup I can guess all the cacheNames (each cache name belongs to a Document/Entity) and could do a "net.​sf.​ehcache.​CacheManager.addCacheIfAbsent(String cacheName)"  to eagerly initialize the CacheManager,  but as said before StageMonitor is already initialized by then.

About the app ... It's a spring-boot / spring-data-mongodb application that uses EhCache, 

Instead of using a conventional org.springframework.cache.ehcache.EhCacheCacheManager we are using a custom one that does the following trick in order to create cache instances at runtime based on ehcache.xml defaultCache configuration.


   ```
 public class CtvEhCacheCacheManager 
    extends      EhCacheCacheManager {
    
    @Override
	protected Cache getMissingCache(String name) {
        Cache parentOne = super.getMissingCache(name);
        if (parentOne == null) {
            Ehcache ehcache = getCacheManager().addCacheIfAbsent(name);            
           if (ehcache != null) {
                return new EhCacheCache(ehcache);
            }
            return null;
        } else {
            return parentOne;
        }
   }
}
```


- ehcache.xml file 

```
    <ehcache xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
	xsi:noNamespaceSchemaLocation="ehcache.xsd" 
	updateCheck="false"
	monitoring="autodetect" 
	dynamicConfig="true">
 
    <sizeOfPolicy maxDepth="200" maxDepthExceededBehavior="abort"/>
    
    <!-- default cache applied per document/entity region without custom config -->
    <defaultCache
           maxEntriesLocalHeap="1500"
           eternal="false"
           timeToLiveSeconds="${ehCache.defaultCache.timeToLiveSeconds}"
           overflowToDisk="false"
           memoryStoreEvictionPolicy="LFU"
           copyOnRead="false"
           copyOnWrite="false"
           transactionalMode="off" 
    />
  <!-- custom definitions goes below -->
</ehcache>

````

**Comment by felixbarny** [comment]
You can extend your custom CacheManager to automatically monitor created caches:

```
 public class CtvEhCacheCacheManager 
    extends      EhCacheCacheManager {

    @Override
    protected Cache getMissingCache(String name) {
        Cache parentOne = super.getMissingCache(name);
        if (parentOne == null) {
            Ehcache ehcache = getCacheManager().addCacheIfAbsent(name);            
           if (ehcache != null) {
                Stagemonitor.getConfiguration(EhCachePlugin.class).monitorCache(chcache);
                return new EhCacheCache(ehcache);
            }
            return null;
        } else {
            return parentOne;
        }
   }
}
```

**Comment by yacota** [comment]
Thanks @felixbarny for your feedback, I wasn't aware of the existence of this method in Stagemonitor class

`Stagemonitor.getConfiguration(EhCachePlugin.class)`

**Comment by yacota** [comment]
Actually you have just made the necessary changes in EhCachePlugin one hour ago :)
Thanks again!

**Comment by felixbarny** [comment]
You're welcome :)
As long as this is not released you could re-implement my added methods.

**Comment by yacota** [comment]
Maybe EhCachePlugin.monitorCache(Cache cache) could check if the received cache is not already registered ?

**Comment by felixbarny** [comment]
That would be a cool enhancement. Would you like to create a pull request for this? You would only have to check whether the metric registry already contains gauges for the Cache.

**Comment by yacota** [comment]
Check https://github.com/stagemonitor/stagemonitor/pull/131

**Comment by yacota** [comment]
Anything wrong with https://github.com/stagemonitor/stagemonitor/pull/132 ? I thought that adding this "relaxed" register method could be helpful to other plugins.
